### PR TITLE
kubeadm: support fetching configuration from the original cluster for 'upgrade diff'

### PR DESCRIPTION
**Pull request by SataQiu**
**What type of PR is this?**
/kind bug
/kind cleanup
/kind feature

**What this PR does / why we need it**:
kubeadm: support fetching configuration from the original cluster for 'upgrade diff'

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm: support fetching configuration from the original cluster for 'upgrade diff'
```

### CODE REVIEW

File: cmd/kubeadm/app/cmd/upgrade/diff.go [hunk 1]
```diff
@@ -75,13 +81,22 @@ func NewCmdDiff(out io.Writer) *cobra.Command {
...
 
 func runDiff(flags *diffFlags, args []string) error {
-
-	// If the version is specified in config file, pick up that value.
-	cfg, err := configutil.LoadInitConfigurationFromFile(flags.cfgPath)
+	var err error
+	var cfg *kubeadmapi.InitConfiguration
+	if flags.cfgPath != "" {
+		cfg, err = configutil.LoadInitConfigurationFromFile(flags.cfgPath)
+	} else {
```
Comment (yastij) [inline cmd/kubeadm/app/cmd/upgrade/diff.go]: can you add a test case for this ?

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 2]
```diff
@@ -47,7 +47,7 @@ func CreateInitStaticPodManifestFiles(manifestDir string, cfg *kubeadmapi.InitCo
 
 // GetStaticPodSpecs returns all staticPodSpecs actualized to the context of the current configuration
 // NB. this methods holds the information about how kubeadm creates static pod manifests.
-func GetStaticPodSpecs(cfg *kubeadmapi.ClusterConfiguration, endpoint *kubeadmapi.APIEndpoint, k8sVersion *version.Version) map[string]v1.Pod {
+func GetStaticPodSpecs(cfg *kubeadmapi.ClusterConfiguration, endpoint *kubeadmapi.APIEndpoint) map[string]v1.Pod {
```
Comment (yastij) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: can you move this refactoring to a separate PR ?

**Review by yastij — COMMENTED** [review]
just nits and missing tests, but lgtm overall

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 2 above)
Comment (neolit123) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: ok, @SataQiu 
please move :
> - Fix the bug that k8sVersion is unused in kubeadm/app/phases/controlplane/manifests.go#getControllerManagerCommand function.
> - manifests.go#GetStaticPodSpecs uses cfg.KubernetesVersion instead of k8sVersion to generate manifests. This PR cleans up redundant k8sVersion parameter.

to separate PR.

Comment (neolit123) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: ideally in other projects i'd just recommend a separate commit, but let's follow what k/k does / recommends.

File: cmd/kubeadm/app/cmd/upgrade/diff.go (hunk 1 above)
Comment (SataQiu) [reply cmd/kubeadm/app/cmd/upgrade/diff.go]: Thanks @yastij
This does lack test cases. But `FetchInitConfigurationFromCluster` requires a running kubernetes cluster. How should we write this kind of test cases? Any examples?

Comment (yastij) [reply cmd/kubeadm/app/cmd/upgrade/diff.go]: If we want to unit test this, we need to mock `FetchInitConfigurationFromCluster`

Comment (neolit123) [reply cmd/kubeadm/app/cmd/upgrade/diff.go]: yeah, let's not do the mock/unit test of the function in question for now.

Comment (yastij) [reply cmd/kubeadm/app/cmd/upgrade/diff.go]: I can settle with that, modulo a TODO or something to make us remember to do this

Comment (SataQiu) [reply cmd/kubeadm/app/cmd/upgrade/diff.go]: @yastij Where should we put the `TODO:`? in `diff_test.go` ?

Comment (yastij) [reply cmd/kubeadm/app/cmd/upgrade/diff.go]: `diff_test.go` seems reasonable to me

Comment (SataQiu) [reply cmd/kubeadm/app/cmd/upgrade/diff.go]: @yastij Done

**Review by neolit123 — COMMENTED** [review]
thanks for the update.
/lgtm
/approve
